### ISSUE TITLE
Bson Serialization in LiteDB not in line with other BSON serializers

### ISSUE THREAD

**Issue by mykemaas** [mbdavid/LiteDB#39]
When serializing BSON, the first four bytes represent the length of the BSON document in total.  This length should include the length bytes.  While the BSON specification is vague on this, popular implementations, including MongoDB include these four bytes in the count.  

For example, assuming our BSON document looks like this (Json representation):

```
{ "FirstName" : "John", "LastName" : "Smith" }
```

The MongoDB BSON serializer will produce the following bytes:

```
2d 00 00 00 02 46 69 72 73 74 4e 61 6d 65 00 05 00 00 00 4a 6f 68 6e 00 02 4c 61
 73 74 4e 61 6d 65 00 06 00 00 00 53 6d 69 74 68 00 00
```

The LiteDB BSON serializer produces this:

```
28 00 00 00 02 46 69 72 73 74 4e 61 6d 65 00 05 00 00 00 4a 6f 68 6e 00 02 4c 61
 73 74 4e 61 6d 65 00 06 00 00 00 53 6d 69 74 68 00 00
```

Notice that all the bytes are the same, except the first one.  2d versus 28.  The MongoDB serializer includes the leading four bytes in the count while LiteDB does not.

The consequence is that the BSON between the two are not compatible, so serialization/deserialization between LiteDB and MongoDB fails when using BSON.

**Comment by mykemaas** [comment]
Quick follow-up... This same bug affects all sub-documents as well.  For example, if I have a document that looks like this:

```
{ "others" : ["Test 1", "Test 2" ] }
```

The array is serialized with a label ("others") and a BSON Document ({ "0" :"Test1", "1":"Test2"), per the BSON spec.  Becuse the length marker is at the document level, both the root document and the document created for the Array have the same length calculation error mentioned above.

**Comment by mykemaas** [comment]
Here is an interim workaround...  simply pass the BSON bytes from MongoDB through this method before deserializing in LiteDB.  This method scans the bytes and adjusts the length markers to be compatible with LiteDB.

```
        /// <summary>
        /// Adjusts the BSON lengths to work around a 
        /// BSON length calculation bug in LiteDB.  Use this
        /// to convert BSON bytes from MongoDB to LiteDB
        /// </summary>
        /// <param name="data">Raw BSON byte array</param>
        static void AdjustBsonLengths(byte[] data)
        {
            bool springLoaded = false;
            for (int i = 0; i < data.Length; i++)
            {
                // If we see a marker for an array or a sub-document, we
                // need to be ready to get the length section once the name
                // string has terminated, hence "spring loaded"
                if (data[i] == 0x04 || data[i] == 0x03)
                {
                    springLoaded = true;
                    i++;
                }

                // If we are at the beginning of the BSON, or we are spring loaded and
                // the name string has terminated, we need to adjust the length bytes by 4
                if (i == 0 || (springLoaded && data[i-1] == 0x00) )
                {
                    int length = BitConverter.ToInt32(data, i);
                    length -= 4;
                    byte[] bytes = BitConverter.GetBytes(length);
                    Array.Copy(bytes, 0, data, i, bytes.Length);
                    i += 3;
                    springLoaded = false;
                }
            }
        }
```

**Comment by mbdavid** [comment]
Hi @mykemaas, thanks for this search and fix. It's quite ease to fix in LiteDB this BSON diference, but the problem is compatibility with older version.

LiteDB has a "file version" on header page to know when file format change. I will use this number to know use current implementation or new implementation. I will fix this in 1.1 versiom.

**Comment by mykemaas** [comment]
Thanks @mbdavid!  This project is seriously amazing.  Very fine work!

**Comment by mbdavid** [comment]
Implemented in new `dev` version 2.0 - commit https://github.com/mbdavid/LiteDB/commit/6d96033a198a75892657dfdae12f0ef16e8609e2
